test/e2e/lifecycle/node_lease.go

```diff
@@ -0,0 +1,164 @@
+				time.Sleep(5 * time.Minute)
+			}
+			if err := framework.WaitForGroupSize(group, int32(framework.TestContext.CloudConfig.NumNodes)); err != nil {
+				framework.Failf("Couldn't restore the original node instance group size: %v", err)
+			}
+
+			if err := framework.WaitForReadyNodes(c, framework.TestContext.CloudConfig.NumNodes, 10*time.Minute); err != nil {
+				framework.Failf("Couldn't restore the original cluster size: %v", err)
+			}
+			// Many e2e tests assume that the cluster is fully healthy before they start.  Wait until
+			// the cluster is restored to health.
+			By("waiting for system pods to successfully restart")
+			err := framework.WaitForPodsRunningReady(c, metav1.NamespaceSystem, systemPodsNo, 0, framework.PodReadyBeforeTimeout, map[string]string{})
+			Expect(err).To(BeNil())
+		})
+
+		It("node lease should be deleted when corresponding node is deleted", func() {
+			leaseClient := c.CoordinationV1beta1().Leases(corev1.NamespaceNodeLease)
+			err := framework.WaitForReadyNodes(c, framework.TestContext.CloudConfig.NumNodes, 10*time.Minute)
+			Expect(err).To(BeNil())
+
+			By("verify node lease exists for every nodes")
+			originalNodes := framework.GetReadySchedulableNodesOrDie(c)
+			Expect(len(originalNodes.Items)).To(Equal(framework.TestContext.CloudConfig.NumNodes))
+
+			Eventually(func() error {
+				pass := true
+				for _, node := range originalNodes.Items {
+					if _, err := leaseClient.Get(node.ObjectMeta.Name, metav1.GetOptions{}); err != nil {
+						framework.Logf("Try to get lease of node %s, but got error: %v", node.ObjectMeta.Name, err)
+						pass = false
+					}
+				}
+				if pass {
+					return nil
+				} else {
+					return fmt.Errorf("some node lease is not ready")
+				}
+			}, 5*time.Minute, 5*time.Second).Should(BeNil())
+
+			targetNumNodes := int32(framework.TestContext.CloudConfig.NumNodes - 1)
+			By(fmt.Sprintf("decreasing cluster size to %d", targetNumNodes))
+			err = framework.ResizeGroup(group, targetNumNodes)
+			Expect(err).To(BeNil())
+			err = framework.WaitForGroupSize(group, targetNumNodes)
+			Expect(err).To(BeNil())
+			err = framework.WaitForReadyNodes(c, framework.TestContext.CloudConfig.NumNodes-1, 10*time.Minute)
+			Expect(err).To(BeNil())
+			targetNodes := framework.GetReadySchedulableNodesOrDie(c)
+			Expect(len(targetNodes.Items)).To(Equal(int(targetNumNodes)))
+
+			By("verify node lease is deleted for the deleted node")
+			var deletedNodeName string
+			for _, originalNode := range originalNodes.Items {
+				originalNodeName := originalNode.ObjectMeta.Name
+				for _, targetNode := range targetNodes.Items {
+					if originalNodeName == targetNode.ObjectMeta.Name {
+						continue
+					}
+				}
+				deletedNodeName = originalNodeName
+				break
+			}
+			Expect(deletedNodeName).NotTo(Equal(""))
+			Eventually(func() error {
+				if _, err := leaseClient.Get(deletedNodeName, metav1.GetOptions{}); err == nil {
+					return fmt.Errorf("node lease is not deleted yet")
```

Comment:
Include the deleted node name in the error message for debugging.

Reply:
Done